diff --git a/llvm/include/llvm/TargetParser/Triple.h b/llvm/include/llvm/TargetParser/Triple.h
index 49ec8de9c528..dc195e8dfcfd 100644
--- a/llvm/include/llvm/TargetParser/Triple.h
+++ b/llvm/include/llvm/TargetParser/Triple.h
@@ -182,7 +182,8 @@ public:
     Mesa,
     SUSE,
     OpenEmbedded,
-    LastVendorType = OpenEmbedded
+    MeeGo,
+    LastVendorType = MeeGo
   };
   enum OSType {
     UnknownOS,
diff --git a/llvm/lib/TargetParser/Triple.cpp b/llvm/lib/TargetParser/Triple.cpp
index 0bbe8a3cedfd..eefd6b90ce2d 100644
--- a/llvm/lib/TargetParser/Triple.cpp
+++ b/llvm/lib/TargetParser/Triple.cpp
@@ -221,6 +221,7 @@ StringRef Triple::getVendorTypeName(VendorType Kind) {
   case Freescale: return "fsl";
   case IBM: return "ibm";
   case ImaginationTechnologies: return "img";
+  case MeeGo: return "meego";
   case Mesa: return "mesa";
   case MipsTechnologies: return "mti";
   case NVIDIA: return "nvidia";
@@ -602,6 +603,7 @@ static Triple::VendorType parseVendor(StringRef VendorName) {
     .Case("mesa", Triple::Mesa)
     .Case("suse", Triple::SUSE)
     .Case("oe", Triple::OpenEmbedded)
+    .Case("meego", Triple::MeeGo)
     .Default(Triple::UnknownVendor);
 }
 
@@ -938,6 +940,9 @@ Triple::Triple(const Twine &Str)
         OS = parseOS(Components[2]);
         if (Components.size() > 3) {
           Environment = parseEnvironment(Components[3]);
+          // MeeGo uses "gnueabi" to mean "gnueabihf"
+          if (Vendor == Triple::MeeGo && Environment == Triple::GNUEABI)
+            Environment = Triple::GNUEABIHF;
           ObjectFormat = parseFormat(Components[3]);
         }
       }
@@ -989,6 +994,10 @@ Triple::Triple(const Twine &ArchStr, const Twine &VendorStr, const Twine &OSStr,
       ObjectFormat(parseFormat(EnvironmentStr.str())) {
   if (ObjectFormat == Triple::UnknownObjectFormat)
     ObjectFormat = getDefaultFormat(*this);
+
+  // MeeGo uses "gnueabi" to mean "gnueabihf"
+  if (Vendor == Triple::MeeGo && Environment == Triple::GNUEABI)
+    Environment = Triple::GNUEABIHF;
 }
 
 std::string Triple::normalize(StringRef Str) {
@@ -1220,6 +1229,9 @@ VersionTuple Triple::getEnvironmentVersion() const {
 StringRef Triple::getEnvironmentVersionString() const {
   StringRef EnvironmentName = getEnvironmentName();
   StringRef EnvironmentTypeName = getEnvironmentTypeName(getEnvironment());
+  // MeeGo uses "gnueabi" to mean "gnueabihf"
+  if (Vendor == Triple::MeeGo && getEnvironment() == Triple::GNUEABIHF)
+    EnvironmentTypeName = "gnueabi";
   EnvironmentName.consume_front(EnvironmentTypeName);
   return EnvironmentName;
 }
diff --git a/llvm/unittests/TargetParser/TripleTest.cpp b/llvm/unittests/TargetParser/TripleTest.cpp
index 4db54a08c0f6..5199caabf224 100644
--- a/llvm/unittests/TargetParser/TripleTest.cpp
+++ b/llvm/unittests/TargetParser/TripleTest.cpp
@@ -642,6 +642,12 @@ TEST(TripleTest, ParsedIDs) {
   EXPECT_EQ(Triple::Linux, T.getOS());
   EXPECT_EQ(Triple::GNUEABI, T.getEnvironment());
 
+  T = Triple("armv7hl-meego-linux-gnueabi");
+  EXPECT_EQ(Triple::arm, T.getArch());
+  EXPECT_EQ(Triple::MeeGo, T.getVendor());
+  EXPECT_EQ(Triple::Linux, T.getOS());
+  EXPECT_EQ(Triple::GNUEABIHF, T.getEnvironment());
+
   T = Triple("i586-pc-haiku");
   EXPECT_EQ(Triple::x86, T.getArch());
   EXPECT_EQ(Triple::PC, T.getVendor());
