diff --git a/clang/lib/Driver/ToolChain.cpp b/clang/lib/Driver/ToolChain.cpp
index 388030592b48..264a849da435 100644
--- a/clang/lib/Driver/ToolChain.cpp
+++ b/clang/lib/Driver/ToolChain.cpp
@@ -424,7 +424,7 @@ std::string ToolChain::getInputFilename(const InputInfo &Input) const {
 
 ToolChain::UnwindTableLevel
 ToolChain::getDefaultUnwindTableLevel(const ArgList &Args) const {
-  return UnwindTableLevel::None;
+  return UnwindTableLevel::Asynchronous;
 }
 
 unsigned ToolChain::GetDefaultDwarfVersion() const {
diff --git a/clang/lib/Driver/ToolChains/Gnu.cpp b/clang/lib/Driver/ToolChains/Gnu.cpp
index 3d0d38972977..b4b35b753a1c 100644
--- a/clang/lib/Driver/ToolChains/Gnu.cpp
+++ b/clang/lib/Driver/ToolChains/Gnu.cpp
@@ -3088,6 +3088,9 @@ void Generic_GCC::printVerboseInfo(raw_ostream &OS) const {
 
 ToolChain::UnwindTableLevel
 Generic_GCC::getDefaultUnwindTableLevel(const ArgList &Args) const {
+  if (getTriple().isMeeGo()) {
+    return UnwindTableLevel::Asynchronous;
+  }
   switch (getArch()) {
   case llvm::Triple::aarch64:
   case llvm::Triple::aarch64_be:
